File: lua/luhn/luhn.lua
return {
  valid = function(s)
      if #s <= 1 then return false end

      local without_spaces = s:gsub("%s+", "")
      local shouldDouble = false
      local sum = 0

      for digit in without_spaces:gmatch("%d") do
          digit = tonumber(digit)
          if digit == nil then return false end

          if shouldDouble then digit = digit * 2 end
          if digit > 9 then digit = digit - 9 end

          sum = sum + digit
          shouldDouble = not shouldDouble
      end

      if sum % 10 == 0 then return true else return false end
  end
}

File: kotlin/luhn/src/main/kotlin/Luhn.kt
object Luhn {

    fun isValid(candidate: String): Boolean {
        val withoutSpaces = candidate.replace(" ", "")

        if (withoutSpaces.length <= 1) return false
        if (withoutSpaces.toBigIntegerOrNull() == null) return false

        val chars = withoutSpaces.reversed().toCharArray()

        val sum =
                chars.foldIndexed(0) { index, acc, char ->
                    var digit = char.digitToInt()

                    if ((index + 1) % 2 == 0) {
                        digit = digit * 2
                        if (digit > 9) digit -= 9
                    }

                    acc + digit
                }

        return sum % 10 == 0
    }
}
